### PR TITLE
feat(torii-grpc): add events subscription

### PR DESCRIPTION
**Pull request by Larkooo**
adds events subscription #2064

### CODE REVIEW

File: crates/torii/grpc/src/server/subscriptions/event.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                };
+
+                if sub.sender.send(Ok(resp)).await.is_err() {
+                    closed_stream.push(*idx);
+                }
+            }
+        }
+
+        for id in closed_stream {
+            trace!(target = LOG_TARGET, id = %id, "Closing entity stream.");
```
Comment (glihm) [inline crates/torii/grpc/src/server/subscriptions/event.rs:109]:
```suggestion
            trace!(target = LOG_TARGET, id = %id, "Closing events stream.");
```

File: crates/torii/grpc/src/server/subscriptions/event.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    type Output = ();
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> std::task::Poll<Self::Output> {
+        let pin = self.get_mut();
+
+        while let Poll::Ready(Some(event)) = pin.simple_broker.poll_next_unpin(cx) {
+            let subs = Arc::clone(&pin.subs_manager);
+            tokio::spawn(async move {
+                if let Err(e) = Service::publish_updates(subs, &event).await {
+                    error!(target = LOG_TARGET, error = %e, "Publishing entity update.");
```
Comment (glihm) [inline crates/torii/grpc/src/server/subscriptions/event.rs:127]:
```suggestion
                    error!(target = LOG_TARGET, error = %e, "Publishing events update.");
```

File: crates/torii/grpc/src/server/mod.rs [hunk 3]
```diff
@@ -895,6 +934,18 @@ impl proto::world::world_server::World for DojoWorld {
 
         Ok(Response::new(events))
     }
+
+    async fn subscribe_events(
+        &self,
+        _request: Request<proto::world::SubscribeEventsRequest>,
```
Comment (glihm) [inline crates/torii/grpc/src/server/mod.rs:940]: Curiosity, why is this not used?

Comment (Larkooo) [reply crates/torii/grpc/src/server/mod.rs:940]: Oh oops, at first th esubscription didnt have a clause so I will add that back

File: crates/torii/grpc/src/server/subscriptions/event.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            .data
+            .trim_end_matches(FELT_DELIMITER)
+            .split(FELT_DELIMITER)
+            .map(FieldElement::from_str)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(ParseError::from)?;
+
+        for (idx, sub) in subs.subscribers.read().await.iter() {
+            // publish all updates if ids is empty or only ids that are subscribed to
+            if sub.keys.is_empty() || keys.starts_with(&sub.keys) {
```
Comment (glihm) [inline crates/torii/grpc/src/server/subscriptions/event.rs:90]: I guess that's the most common operation on regular event, we just want a sub-array from the beginning. We may update this later if needed.

We may document the behavior though, to ensure the user can understand the clause. I'll add an issue on the dojo-book to ensure that we describe better what's available and how. :)

Thanks for the work done here mate!

**Review by glihm — APPROVED** [review]

**Pull request merged**